src/main/java/digital/slovensko/autogram/core/Autogram.java

```diff
@@ -89,22 +90,37 @@ public void startVisualization(SigningJob job) {
         });
     }
 
+    private void signCommon(SigningJob job, SigningKey signingKey, Consumer<SigningJob> callback) {
+        try {
+            job.signWithKeyAndRespond(signingKey);
+            if (batch == null || batch.isEnded() || batch.isAllProcessed())
+                passwordManager.reset();
+
+            callback.accept(job);
+        } catch (PINIncorrectException e) {
+            passwordManager.reset();
+            throw e;
+        } catch (AutogramException e) {
+            throw e;
+        } catch (DSSException e) {
+            throw AutogramException.createFromDSSException(e);
+        } catch (IllegalArgumentException e) {
+            throw AutogramException.createFromIllegalArgumentException(e);
+        } catch (Exception e) {
+            throw new UnrecognizedException(e);
+        }
+    }
+
     public void sign(SigningJob job, SigningKey signingKey) {
         ui.onWorkThreadDo(() -> {
             try {
-                job.signWithKeyAndRespond(signingKey);
-                if (batch == null || batch.isEnded()) passwordManager.reset();
-                ui.onUIThreadDo(() -> ui.onSigningSuccess(job));
+                signCommon(job, signingKey, (jobNew) -> {
```

Comment:
Drzal by som tu konvenciu `...AndThen` ak to ocakava ze na uspesnom konci to zavola ten callback.